README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

Data Table package makes it possible to create a data table with hundreds
of rows, which is not possible using the default ```DataTable``` widget
because of performance reasons.

## Features
- ScalableDataTable only builds the rows that are visible at each moment
(similar to ListView.builder)
- Horizontal scroll if , use the ```minWidth``` property.
- Sticky Header: The header will be visible at all times and the rows will
get the rest of the available space.

## Getting started
Add the package to your project by running ```flutter pub add scalable_data_table```.
You can also add it manually in the ```pubspec.yaml``` file.

## Usage

Check the `/example` folder for the example project.

```dart
class UsersTable extends StatelessWidget {
  const UsersTable({Key? key}) : super(key: key);

  static final _dateFormat = DateFormat('HH:mm dd/MM');

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<List<User>>(
      future: createUsers(),
      builder: (context, snapshot) => ScalableDataTable(
        header: DefaultTextStyle(
          style: TextStyle(
            fontSize: 16,
            fontWeight: FontWeight.bold,
            color: Colors.grey[700],
          ),
          child: ScalableTableHeader(
            columnWrapper: columnWrapper,
            children: [
              const SizedBox(),
              const Text('Created At'),
              const Text('Name'),
              const Text('Surname'),
              const Text('Points'),
              const Text('Interests'),
            ],
          ),
        ),
        rowBuilder: (context, index) {
          final user = snapshot.data![index];
          return ScalableTableRow(
            columnWrapper: columnWrapper,
            color: MaterialStateColor.resolveWith((states) =>
                (index % 2 == 0) ? Colors.grey[200]! : Colors.transparent),
            children: [
              Align(
                alignment: Alignment.centerRight,
                child: Text('${user.index}.'),
              ),
              Text(_dateFormat.format(user.createdAt)),
              Text(user.name),
              Text(user.surname),
              Text('${user.points} pts'),
              Text(user.interests.join(', '), maxLines: 2),
            ],
          );
        },
        emptyBuilder: (context) => const Text('No users yet...'),
        itemCount: snapshot.data?.length ?? -1,
        minWidth: 1000, // max(MediaQuery.of(context).size.width, 1000),
        textStyle: TextStyle(color: Colors.grey[700], fontSize: 14),
      ),
    );
  }

  Widget columnWrapper(BuildContext context, int columnIndex, Widget child) {
    const padding = EdgeInsets.symmetric(horizontal: 10);
    switch (columnIndex) {
      case 0:
        return Container(
          width: 60,
          padding: padding,
          child: child,
        );
      case 1:
        return Container(
          width: 100,
          padding: padding,
          child: child,
        );
      case 5:
        return Expanded(
          flex: 3,
          child: Container(
            padding: padding,
            child: child,
          ),
        );
      default:
        return Expanded(
          child: Container(
            padding: padding,
            child: child,
          ),
        );
    }
  }
}
```

![example gif](https://raw.githubusercontent.com/LucaDillenburg/Scalable-Data-Table/stable/assets/example.gif)


CHANGELOG.md
## 0.0.1

* TODO: First version of ScalableDataTable, ScalableTableHeader and ScalableTableRow.


example/lib/data.dart
import 'dart:math';

class User {
  final int index;
  final String name;
  final String surname;
  final int points;
  final DateTime createdAt;
  final List<String> interests;

  User({
    required this.index,
    required this.name,
    required this.surname,
    required this.points,
    required this.createdAt,
    required this.interests,
  });
}

const _names = [
  'Oliver',
  'Jake',
  'Noah',
  'James',
  'Jack',
  'Connor',
  'Liam',
  'John',
  'Harry',
  'Callum',
  'Mason',
  'Robert',
  'Michael',
  'Charlie',
  'Kyle',
  'William',
  'Thomas',
  'Joe',
  'Ethan',
  'David',
  'George',
  'Reece',
  'Michael',
  'Richard',
  'Oscar',
  'Alexander',
  'Joseph',
  'James',
  'Charlie',
  'James',
  'Charles',
  'William',
  'Damian',
  'Daniel',
  'Thomas',
];

const _surnames = [
  'Smith',
  'Murphy',
  'Smith',
  'Jones',
  'O\'Kelly',
  'Johnson',
  'Williams',
  'O\'Sullivan',
  'Williams',
  'Brown',
  'Walsh',
  'Brown',
  'Taylor',
  'Smith',
  'Jones',
  'Davies',
  'O\'Brien',
  'Miller',
  'Wilson',
  'Byrne',
  'Davis',
  'Evans',
  'O\'Ryan',
  'Garcia',
  'Thomas',
  'O\'Connor',
  'Rodriguez',
  'Roberts',
  'O\'Neill',
  'Wilson',
];

final _random = Random();
String _createName() => _names[_random.nextInt(_names.length)];
String _createSurname() => _surnames[_random.nextInt(_surnames.length)];

Future<List<User>> createUsers() {
  return Future.delayed(
    const Duration(seconds: 3),
    () => List.generate(
      500,
      (index) => User(
        index: index,
        name: _createName(),
        surname: _createSurname(),
        createdAt: DateTime.now().subtract(Duration(minutes: index * 30)),
        points: index * 5,
        interests: List.generate(
          _random.nextInt(30),
          (_) => '${_createName()} ${_createSurname()}',
        ),
      ),
    ),
  );
}


example/lib/table.dart
import 'package:example/data.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:scalable_data_table/scalable_data_table.dart';
import 'package:intl/intl.dart';

class UsersTable extends StatelessWidget {
  const UsersTable({Key? key}) : super(key: key);

  static final _dateFormat = DateFormat('HH:mm dd/MM');

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<List<User>>(
      future: createUsers(),
      builder: (context, snapshot) => ScalableDataTable(
        header: DefaultTextStyle(
          style: TextStyle(
            fontSize: 16,
            fontWeight: FontWeight.bold,
            color: Colors.grey[700],
          ),
          child: ScalableTableHeader(
            columnWrapper: columnWrapper,
            children: [
              const SizedBox(),
              const Text('Created At'),
              const Text('Name'),
              const Text('Surname'),
              const Text('Points'),
              const Text('Interests'),
            ],
          ),
        ),
        rowBuilder: (context, index) {
          final user = snapshot.data![index];
          return ScalableTableRow(
            columnWrapper: columnWrapper,
            color: MaterialStateColor.resolveWith((states) =>
                (index % 2 == 0) ? Colors.grey[200]! : Colors.transparent),
            children: [
              Align(
                alignment: Alignment.centerRight,
                child: Text('${user.index}.'),
              ),
              Text(_dateFormat.format(user.createdAt)),
              Text(user.name),
              Text(user.surname),
              Text('${user.points} pts'),
              Text(user.interests.join(', '), maxLines: 2),
            ],
          );
        },
        emptyBuilder: (context) => const Text('No users yet...'),
        itemCount: snapshot.data?.length ?? -1,
        minWidth: 1000, // max(MediaQuery.of(context).size.width, 1000),
        textStyle: TextStyle(color: Colors.grey[700], fontSize: 14),
      ),
    );
  }

  Widget columnWrapper(BuildContext context, int columnIndex, Widget child) {
    const padding = EdgeInsets.symmetric(horizontal: 10);
    switch (columnIndex) {
      case 0:
        return Container(
          width: 60,
          padding: padding,
          child: child,
        );
      case 1:
        return Container(
          width: 100,
          padding: padding,
          child: child,
        );
      case 5:
        return Expanded(
          flex: 3,
          child: Container(
            padding: padding,
            child: child,
          ),
        );
      default:
        return Expanded(
          child: Container(
            padding: padding,
            child: child,
          ),
        );
    }
  }
}


lib/src/header_row.dart
import 'package:collection/collection.dart';
import 'package:flutter/material.dart';
import 'package:flutter/painting.dart';
import 'package:flutter/widgets.dart';

class ScalableTableHeader extends StatelessWidget {
  final List<Widget> children;

  /// Use [Container] width or [Expanded] with width wrapper
  final Widget Function(BuildContext, int columnIndex, Widget child)
      columnWrapper;

  final EdgeInsets? padding;

  const ScalableTableHeader({
    required this.children,
    required this.columnWrapper,
    this.padding,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return _HorizontalTable(
      children: children,
      columnWrapper: columnWrapper,
      padding: padding,
    );
  }
}

class ScalableTableRow extends StatelessWidget {
  final List<Widget> children;

  final MaterialStateColor? color;

  /// Use [Container] width or [Expanded] with width wrapper
  final Widget Function(BuildContext, int columnIndex, Widget child)
      columnWrapper;

  final Function()? onTap;

  final EdgeInsets? padding;

  const ScalableTableRow({
    required this.children,
    required this.columnWrapper,
    required this.color,
    this.onTap,
    this.padding,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final child = GestureDetector(
      onTap: onTap,
      child: Container(
        color: color ?? Colors.transparent,
        child: _HorizontalTable(
          children: children,
          columnWrapper: columnWrapper,
          padding: padding,
        ),
      ),
    );

    if (onTap == null) return child;
    return MouseRegion(
      cursor: SystemMouseCursors.click,
      child: child,
    );
  }
}

class _HorizontalTable extends StatelessWidget {
  final List<Widget> children;

  /// Use [Container] width or [Expanded] with width wrapper
  final Widget Function(BuildContext, int columnIndex, Widget child)
      columnWrapper;

  final EdgeInsets? padding;

  const _HorizontalTable({
    required this.children,
    required this.columnWrapper,
    this.padding,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: (padding ?? const EdgeInsets.symmetric(horizontal: 20)) +
          const EdgeInsets.symmetric(vertical: 4),
      child: Row(
        mainAxisSize: MainAxisSize.max,
        children: children
            .mapIndexed((index, child) => columnWrapper(context, index, child))
            .toList(),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:example/data.dart';
import 'package:scalable_data_table/scalable_data_table.dart';
import 'package:intl/intl.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: const MyHomePage(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const Padding(
            padding: EdgeInsets.all(8.0),
            child: Text(
              'Scalable Data Table',
              style: TextStyle(fontSize: 25, fontWeight: FontWeight.bold),
            ),
          ),
          const Expanded(child: UsersTable()),
        ],
      ),
    );
  }
}

class UsersTable extends StatelessWidget {
  const UsersTable({Key? key}) : super(key: key);

  static final _dateFormat = DateFormat('HH:mm dd/MM');

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<List<User>>(
      future: createUsers(),
      builder: (context, snapshot) => ScalableDataTable(
        header: DefaultTextStyle(
          style: TextStyle(
            fontSize: 16,
            fontWeight: FontWeight.bold,
            color: Colors.grey[700],
          ),
          child: ScalableTableHeader(
            columnWrapper: columnWrapper,
            children: [
              const SizedBox(),
              const Text('Created At'),
              const Text('Name'),
              const Text('Surname'),
              const Text('Points'),
              const Text('Interests'),
            ],
          ),
        ),
        rowBuilder: (context, index) {
          final user = snapshot.data![index];
          return ScalableTableRow(
            columnWrapper: columnWrapper,
            color: MaterialStateColor.resolveWith((states) =>
                (index % 2 == 0) ? Colors.grey[200]! : Colors.transparent),
            children: [
              Align(
                alignment: Alignment.centerRight,
                child: Text('${user.index}.'),
              ),
              Text(_dateFormat.format(user.createdAt)),
              Text(user.name),
              Text(user.surname),
              Text('${user.points} pts'),
              Text(user.interests.join(', '), maxLines: 2),
            ],
          );
        },
        emptyBuilder: (context) => const Text('No users yet...'),
        itemCount: snapshot.data?.length ?? -1,
        minWidth: 1000, // max(MediaQuery.of(context).size.width, 1000),
        textStyle: TextStyle(color: Colors.grey[700], fontSize: 14),
      ),
    );
  }

  Widget columnWrapper(BuildContext context, int columnIndex, Widget child) {
    const padding = EdgeInsets.symmetric(horizontal: 10);
    switch (columnIndex) {
      case 0:
        return Container(
          width: 60,
          padding: padding,
          child: child,
        );
      case 1:
        return Container(
          width: 100,
          padding: padding,
          child: child,
        );
      case 5:
        return Expanded(
          flex: 3,
          child: Container(
            padding: padding,
            child: child,
          ),
        );
      default:
        return Expanded(
          child: Container(
            padding: padding,
            child: child,
          ),
        );
    }
  }
}
